_posts/2017-07-29-ruby-oop.md

```diff
@@ -0,0 +1,383 @@
+
+We can now make a new `Point` like so:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.inspect #=> #<Point:0x2ae37f0 @x=1.5, @y=-3.0>
+```
+
+The attributes we've defined (`@x`, `@y`) can only be accessed inside of the class's own methods. We're going to need to use these attributes of a point in order to do our calculations. To do so, we can define methods that will return the values of our attributes:
+
+```ruby
+# A location in 2D space.
+class Point
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+
+  def x
+    @x
+  end
+
+  def y
+    @y
+  end
+end
+```
+Since this is such a common thing in Ruby, the language has shortcuts to quickly define such simple methods that reflect an objects attributes:
+
+- `attr_reader :var` - Creates a read-only method for attribute `@var`.
+- `attr_writer :var` - Creates a write-only method for attribute `@var`.
+- `attr_accessor :var` - Creates both kinds of methods to read and write to this attribute.
+
+So, our `Point` can be described simply as:
+```ruby
+# A location in 2D space.
+class Point
+  attr_reader :x
+
+  attr_reader :y
+
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+end
+```
+> In our case, we will only need to ever read these attributes, so I just used `attr_reader`. `attr_accessor` would have also worked here.
+
+Playing with our new readers:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.x #=> 1.5
+puts point.y #=> -3.0
+```
+> You can read more about attributes, readers, setters, and more [here](http://ruby-for-beginners.rubymonstas.org/writing_classes/instance_variables.html).
+
+What else would be helpful to do with our point, or perhaps any other point? Well, with any two points, you can calculate the distance between them.
+
+Ruby has a rich standard library, which includes some common mathematical functions. We need the pythagorean theorem here to calculate the distance between two points, and if we do some searching we can see Ruby already provides a method for us to calculate this instead of having to implement it ourselves:
+
+[https://ruby-doc.org/core-2.4.0/Math.html#method-c-hypot](https://ruby-doc.org/core-2.4.0/Math.html#method-c-hypot)
+```ruby
+Math.hypot(3, 4) #=> 5.0
+```
+
+Let's implement it to our `Point`. (Note: I've omitted the rest of the class we already wrote to save space, ad nauseum)
+```ruby
+# A location in 2D space.
+class Point
+  # Calculates the distance to another Point object
+  def distance(other)
+    x_difference = x - other.x
+    y_difference = y - other.y
+
+    Math.hypot x_difference, y_difference
+  end
+end
+```
+> Here, the `x` in `x_difference = x - other.x` is actually calling this objects *own* `attr_reader` method internally that fetches its `@x` attribute. In another language, you may write something like `self.x` to refer to this value. You can do that here too in Ruby, but in most cases you can omit `self.`, because Ruby understands what you mean.
```

Comment:
By saying "actually" here I'm attempting to distinguish it as a method call, with the assumption the reader might think it's just another temporary variable / ivar.

Beyond this I'd rather not dive into the tangent of how other language's method / function scopes work.